scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,416 @@
+import math
+import heapq
+import functools
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import numpy as np
+
+
+class CubatureRule(ABC):
+    def __init__(self, nodes, weights):
+        self.nodes = nodes
+        self.weights = weights
+
+    def dimension(self):
+        """
+        Return the input dimension that the cubature rule is expecting `f` to have, e.g.
+        for a cubature rule for triple integrals, this would be three.
+        """
+        return self.nodes.shape[-1]
+
+    @abstractmethod
+    def estimate(self, f, a, b, args):
+        """
+        Calculate estimate of integral of f in region described by corners a and b.
+
+        f should accept arrays of shape (input_dim, num_eval_points) and return arrays
+        of shape (output_dim_1, ..., output_dim_n, num_eval_points).
+
+        Should return (integral_estimate, error_estimate).
+        """
+        pass
+
+
+class CubatureRuleLinearError(CubatureRule):
+    def __init__(self, nodes, weights, error_nodes, error_weights):
+        super().__init__(nodes, weights)
+
+        self.error_nodes = error_nodes
+        self.error_weights = error_weights
+
+    def estimate(self, f, a, b, args):
+        # The underlying cubature rule is presumed to be for the hypercube [0, 1]^n.
+        #
+        # To handle arbitrary regions of integration, it's necessary to apply a linear
+        # change of coordinates to map each interval [a[i], b[i]] to [-1, 1].
+        a = a[:, np.newaxis]
+        b = b[:, np.newaxis]
+
+        lengths = b - a
+
+        # nodes have shape (input_dim, eval_points)
+        nodes = (self.nodes + 1) * (lengths / 2) + a
+        error_nodes = (self.error_nodes + 1) * (lengths / 2) + a
+
+        # Also need to multiply the weights by a scale factor equal to the determinant
+        # of the Jacobian for this coordinate change.
+        weight_scale_factor = math.prod(lengths / 2)
+
+        # weights have shape (eval_points,)
+        weights = self.weights * weight_scale_factor
+        error_weights = self.error_weights * weight_scale_factor
+
+        # f(nodes) will have shape (eval_points, output_dim)
+        # integral_estimate should have shape (output_dim)
+        integral_estimate = np.sum(weights * f(nodes, *args), axis=-1)
+
+        error_estimate = abs(np.sum(
```

Comment:
Should use `np.abs` here (and throughout). This is so we can generalise to `xp.abs` when we add support for the array API.